oprf/oprf.go

```diff
@@ -0,0 +1,229 @@
+package oprf
+
+import (
+	"crypto/subtle"
+	"errors"
+	"math/big"
+
+	"github.com/cloudflare/circl/oprf/group"
+)
+
+const (
+	// OPRFP256 is the constant to represent the OPRF P-256 with SHA-512 (SSWU-RO) group.
+	OPRFP256 uint16 = 0x0003
+	// OPRFP384 is the constant to represent the OPRF P-384 with SHA-512 (SSWU-RO) group.
+	OPRFP384 uint16 = 0x0004
+	// OPRFP521 is the constant to represent the OPRF P-521 with SHA-512 (SSWU-RO) group.
+	OPRFP521 uint16 = 0x0005
+)
+
+var (
+	// OPRFMode is the context string to define a OPRF.
+	OPRFMode *big.Int = big.NewInt(0)
+)
+
+var (
+	// ErrUnsupportedGroup is an error stating that the ciphersuite chosen is not supported
+	ErrUnsupportedGroup = errors.New("the chosen group is not supported")
+)
+
+// BlindToken corresponds to a token that has been blinded.
+// Internally, it is a serialized Element.
+type BlindToken []byte
+
+// IssuedToken corresponds to a token that has been issued.
+// Internally, it is a serialized Element.
+type IssuedToken []byte
+
+// Token is the object issuance of the protocol.
+type Token struct {
+	data  []byte
+	blind *group.Scalar
+}
+
+// Evaluation corresponds to the evaluation over a token.
+type Evaluation struct {
+	element []byte
+}
+
+// KeyPair is an struct containing a public and private key.
+type KeyPair struct {
+	PubK  *group.Element
+	PrivK *group.Scalar
+}
+
+// Client is a representation of a Client during protocol execution.
+type Client struct {
+	suite *group.Ciphersuite
+	ctx   []byte
+}
+
+// ClientContext implements the functionality of a Client.
+type ClientContext interface {
+	// Request generates a token and blinded data.
+	Request(in []byte) (*Token, *BlindToken, error)
```

Comment:
Why is this a function on `ClientContext`, instead of `Client`?

Reply:
Mmm.. Client is the struct, which has the method. But this is the interface that defines the methods that can be used by other objects implementing it. But, I'll remove the interface all together ;)